### PR TITLE
Use multilib build for x265 for 10/12bit access

### PR DESCRIPTION
**Pull request by phpb-com**
Please consider using multilib build of x265 codec to allow for processing of 10+ bit videos.

### CODE REVIEW

**Comment by wader** [comment]
Hey, thanks for the PR. I didn't know about the multilib thing for x265, and yes that looks like something we use. Hopefully have time to look at it today or during the weekend.

**Comment by wader** [comment]
> This actually might be broken on multi-platform, but I would appreciate any help to fix it.

Did you notice it failing on arm64? can't see what in multilib.sh could make it fail, but yeah let's try build and see. I don't own any arm machines at the moment so a bit more cumbersome to get a test env.

**Comment by wader** [comment]
Actually remember now that we can try multi platform by pushing a tag with your changes. Here is the build running https://github.com/wader/static-ffmpeg/actions/runs/3066990488 let's see how it goes

File: Dockerfile [hunk 1]
```diff
@@ -666,17 +666,11 @@ RUN \
...
-    -DENABLE_AGGRESSIVE_CHECKS=ON \
-    -DCMAKE_ASM_NASM_FLAGS=-w-macro-params-legacy \
-    -DENABLE_CLI=OFF \
-    -DENABLE_NASM=ON \
-    -DCMAKE_BUILD_TYPE=Release \
-    ../../source && \
-  make -j$(nproc) install
+  sed -i '/^cmake / s/$/ -DCMAKE_VERBOSE_MAKEFILE=ON -DENABLE_AGGRESSIVE_CHECKS=ON -DCMAKE_ASM_NASM_FLAGS=-w-macro-params-legacy -DENABLE_NASM=ON -DCMAKE_BUILD_TYPE=Release/' ./multilib.sh && \
+  sed -i 's/ -DENABLE_SHARED=OFF//g' ./multilib.sh && \
+  sed -i '/^cmake / s/$/ -DENABLE_SHARED=OFF/' ./multilib.sh && \
```
Comment (wader) [inline Dockerfile:671]: My sed skills is not what they should be. This add flags last to cmake lines and then moves `-DENABLE_SHARED=OFF` last?

Wonder if it is possible to pass arguments or defines via env? another alternative  is to use sed to add a `${CMAKEFLAGS}` similar to MAKEFLAGS and use that?

**Comment by wader** [comment]
Looks like CI build succeeded (amd64) but the arm64 multiplatform build failed, so that has to be look into:

```
....
[100%] Linking CXX executable x265
/usr/bin/cmake -E cmake_link_script CMakeFiles/cli.dir/link.txt --verbose=1
/usr/bin/c++ -O3 -s -static-libgcc -fno-strict-overflow -fstack-protector-all -fPIC -O3 -DNDEBUG -Wl,-z,relro,-z,now -rdynamic -L. CMakeFiles/cli.dir/input/input.cpp.o CMakeFiles/cli.dir/input/y4m.cpp.o CMakeFiles/cli.dir/input/yuv.cpp.o CMakeFiles/cli.dir/output/output.cpp.o CMakeFiles/cli.dir/output/raw.cpp.o CMakeFiles/cli.dir/output/reconplay.cpp.o CMakeFiles/cli.dir/output/y4m.cpp.o CMakeFiles/cli.dir/output/yuv.cpp.o CMakeFiles/cli.dir/x265.cpp.o CMakeFiles/cli.dir/x265cli.cpp.o CMakeFiles/cli.dir/abrEncApp.cpp.o -o x265  libx265.a -lpthread -lrt -ldl -lnuma -Wl,-Bstatic -lx265_main10 -lx265_main12 -Wl,-Bdynamic 
/usr/lib/gcc/aarch64-alpine-linux-musl/11.2.1/../../../../aarch64-alpine-linux-musl/bin/ld: cannot find -lx265_main10: No such file or directory
/usr/lib/gcc/aarch64-alpine-linux-musl/11.2.1/../../../../aarch64-alpine-linux-musl/bin/ld: cannot find -lx265_main12: No such file or directory
collect2: error: ld returned 1 exit status
make[2]: *** [CMakeFiles/cli.dir/build.make:261: x265] Error 1
make[2]: Leaving directory '/x265_3.5/build/linux/8bit'
make[1]: *** [CMakeFiles/Makefile2:204: CMakeFiles/cli.dir/all] Error 2
make[1]: Leaving directory '/x265_3.5/build/linux/8bit'
make: *** [Makefile:139: all] Error 2
ar: libx265_main10.a: No such file or directory
The command '/bin/sh -c wget $WGET_OPTS -O x265.tar.bz2 "$X265_URL" &&   echo "$X265_SHA256  x265.tar.bz2" | sha256sum --status -c - &&   tar xf x265.tar.bz2 &&   cd x265_*/build/linux &&   sed -i '/^cmake / s/$/ -DCMAKE_VERBOSE_MAKEFILE=ON -DENABLE_AGGRESSIVE_CHECKS=ON -DCMAKE_ASM_NASM_FLAGS=-w-macro-params-legacy -DENABLE_NASM=ON -DCMAKE_BUILD_TYPE=Release/' ./multilib.sh &&   sed -i 's/ -DENABLE_SHARED=OFF//g' ./multilib.sh &&   sed -i '/^cmake / s/$/ -DENABLE_SHARED=OFF/' ./multilib.sh &&   MAKEFLAGS="-j$(nproc)" ./multilib.sh &&   make -C 8bit -j$(nproc) install' returned a non-zero code: 9
```

**Comment by wader** [comment]
There is a bunch of aarch64 and 10+ bit depth issues that might be a good start, ex https://bitbucket.org/multicoreware/x265_git/issues/604/linux-arm-aarch64-build-failing-for-high

**Comment by wader** [comment]
Great, i will probably digg some also if i get time. Let's sync findings here

**Comment by wader** [comment]
Thanks for testing and researching. It sounds like it might be best to just use master then? that should fix all current build issues we see? and it should be fine auto version update-wise also as it's not updated very frequently at least.

So maybe do something similar how x264 is cloned then?

I wonder why x265 haven't done any stable releases in a while

File: Dockerfile [hunk 2]
```diff
@@ -657,19 +657,24 @@ RUN \
...
   tar xf x265.tar.bz2 && \
-  cd x265_*/build/linux && \
-  sed -i '/^cmake / s/$/ -DCMAKE_VERBOSE_MAKEFILE=ON -DENABLE_AGGRESSIVE_CHECKS=ON -DCMAKE_ASM_NASM_FLAGS=-w-macro-params-legacy -DENABLE_NASM=ON -DCMAKE_BUILD_TYPE=Release/' ./multilib.sh && \
+  cd multicoreware-x265_git-*/build/linux && \
+  sed -i '/^cmake / s/$/ -G "Unix Makefiles" ${CMAKEFLAGS}/' ./multilib.sh && \
   sed -i 's/ -DENABLE_SHARED=OFF//g' ./multilib.sh && \
-  sed -i '/^cmake / s/$/ -DENABLE_SHARED=OFF/' ./multilib.sh && \
-  MAKEFLAGS="-j$(nproc)" ./multilib.sh && \
+  MAKEFLAGS="-j$(nproc)" \
+  CMAKEFLAGS="-DENABLE_SHARED=OFF -DCMAKE_VERBOSE_MAKEFILE=ON -DENABLE_AGGRESSIVE_CHECKS=ON -DCMAKE_ASM_NASM_FLAGS=-w-macro-params-legacy -DENABLE_NASM=ON -DCMAKE_BUILD_TYPE=Release" \
```
[the remark this answers is not in the document]
Comment (wader) [reply Dockerfile:676]: Yeap!

**Comment by wader** [comment]
> Excellent, I'll get on it and produce a clean commit (rewriting the previous ones to have a better git history). As for the release cut from x265, I am not sure but that is not really a bit issue since bunch of projects use master already.

Ok good, i think rather stabelish master then 3.5 + random patches :)

> I am trying to do some HDR and non HDR encodings to make sure there is not anything funny with master branch. Once that all done, I'll get to cleanning of the commit into this branch. Thanks!

Yes some testing would be great 👍  if you come up with some quick example encoding that can be done to verify that it works you can add it to the sanity tests if you want.

**Comment by wader** [comment]
> Ok, latest commit should be it, unless I made a bunch of mistakes in the "bump" section.

No worries, i can help fix the bump stuff

**Comment by wader** [comment]
> Tried building on C7G and cannot solve the `ERROR: x265 not found using pkg-config` problem. Works as expected on amd64 platform. `x265.pc` is present with the rest of the others and its content seems to be correct:

`ERROR: x265 not found using pkg-config` is when running ffmpeg configure? if so we should be able to see more details in `config.log` and it should have dump it to stdout on error.

I triggered a new build using your branch https://github.com/wader/static-ffmpeg/actions/runs/3072459017

File: Dockerfile [hunk 3]
```diff
@@ -653,30 +653,24 @@ RUN \
   ./configure --enable-pic --enable-static --disable-cli --disable-lavf --disable-swscale && \
   make -j$(nproc) install
 
-# bump: x265 /X265_VERSION=([\d.]+)/ https://bitbucket.org/multicoreware/x265_git.git|^3
+# x265 release is over 1 years old and master branch has a lot of fixes and improvements, so we checkout commit so no hash is needed
```
[the remark this answers is not in the document]
Comment (wader) [reply Dockerfile:656]: Nice with comment. And bump stuff we will figure out later

**Comment by wader** [comment]
Fails because of ”read-only segment has dynamic relocations”. I guess probably because of some mix of ro and pic flags (used for hardening so maybe a bit unusal?) not being used/respected somewhere. Next step maybe figure where that segment origintes from? Or maybe known x266 issue? (Will check later, on the phone)

File: Dockerfile [hunk 4]
```diff
@@ -653,30 +653,24 @@ RUN \
...
-  make -j$(nproc) install
+  wget $WGET_OPTS -O x265_git.tar.bz2 "$X265_URL" && \
+  tar xf x265_git.tar.bz2 && \
+  cd multicoreware-x265_git-*/build/linux && \
+  sed -i '/^cmake / s/$/ -G "Unix Makefiles" ${CMAKEFLAGS}/' ./multilib.sh && \
+  sed -i 's/ -DENABLE_SHARED=OFF//g' ./multilib.sh && \
+  MAKEFLAGS="-j$(nproc)" \
+  CMAKEFLAGS="-DENABLE_SHARED=OFF -DCMAKE_VERBOSE_MAKEFILE=ON -DENABLE_AGGRESSIVE_CHECKS=ON -DCMAKE_ASM_NASM_FLAGS=-w-macro-params-legacy -DENABLE_NASM=ON -DCMAKE_BUILD_TYPE=Release" \
+  ./multilib.sh && \
+  make -C 8bit -j$(nproc) install
```
Comment (wader) [inline Dockerfile:673]: `-C 8bit` needed?

**Comment by wader** [comment]
> I did the encoding and things look "sane", but not sure how I would go about adding it. I could probably snap a quick iPhone video which would weight at few tens of MiBs, but encoding that would take some CPU time and wall-clock time. If you are OK with me just eyeballing it, consider it done.

Some ffmpeg commands that produce and probe would be ideally, so something like this:

```sh
$ ffmpeg -y -v quiet -f lavfi -i testsrc -pix_fmt yuv444p12le -t 20ms -c:v libx265 test.mp4
x265 [info]: HEVC encoder version 3.4+31-6722fce1f
x265 [info]: build info [Mac OS X][clang 12.0.0][64 bit] 12bit
x265 [info]: using cpu capabilities: MMX2 SSE2Fast LZCNT SSSE3 SSE4.2 AVX FMA3 BMI2 AVX2
x265 [warning]: halving the quality when psy-rd is enabled for 444 input. Setting cbQpOffset = 6 and crQpOffset = 6
x265 [info]: Main 4:4:4 12 profile, Level-2 (Main tier)
x265 [info]: Thread pool created using 4 threads
x265 [info]: Slices                              : 1
x265 [info]: frame threads / pool features       : 2 / wpp(4 rows)
x265 [warning]: Source height < 720p; disabling lookahead-slices
x265 [info]: Coding QT: max CU size, min CU size : 64 / 8
x265 [info]: Residual QT: max TU size, max depth : 32 / 1 inter / 1 intra
x265 [info]: ME / range / subpel / merge         : hex / 57 / 2 / 3
x265 [info]: Keyframe min / max / scenecut / bias  : 25 / 250 / 40 / 5.00
x265 [info]: Cb/Cr QP Offset                     : 6 / 6
x265 [info]: Lookahead / bframes / badapt        : 20 / 4 / 2
x265 [info]: b-pyramid / weightp / weightb       : 1 / 1 / 0
x265 [info]: References / ref-limit  cu / depth  : 3 / off / on
x265 [info]: AQ: mode / str / qg-size / cu-tree  : 2 / 1.0 / 32 / 1
x265 [info]: Rate Control / qCompress            : CRF-28.0 / 0.60
x265 [info]: tools: rd=3 psy-rd=2.00 early-skip rskip mode=1 signhide tmvp
x265 [info]: tools: b-intra strong-intra-smoothing deblock sao
x265 [info]: frame I:      1, Avg QP:32.75  kb/s: 324.00
x265 [info]: consecutive B-frames: 100.0% 0.0% 0.0% 0.0% 0.0%

encoded 1 frames in 0.06s (17.32 fps), 324.00 kb/s, Avg QP:32.75

$ ffprobe -v quiet -of json -show_streams -i test.mp4 | jq -r '.streams[0].pix_fmt'
yuv444p12le
```

Would that be enough to see if libx265 didn't fail and ffmpeg didn't insert some transformation because the codec didn't support to pixel format? im not that familiar with HDR, does it mean something more than 10+ bit depth? color spaces etc?

This is all metadata ffprobe can see about the above stream:
```json
{
    "streams": [
        {
            "index": 0,
            "codec_name": "hevc",
            "codec_long_name": "H.265 / HEVC (High Efficiency Video Coding)",
            "profile": "Rext",
            "codec_type": "video",
            "codec_tag_string": "hev1",
            "codec_tag": "0x31766568",
            "width": 320,
            "height": 240,
            "coded_width": 320,
            "coded_height": 240,
            "closed_captions": 0,
            "film_grain": 0,
            "has_b_frames": 2,
            "sample_aspect_ratio": "1:1",
            "display_aspect_ratio": "4:3",
            "pix_fmt": "yuv444p12le",
            "level": 60,
            "color_range": "tv",
            "field_order": "progressive",
            "refs": 1,
            "id": "0x1",
            "r_frame_rate": "25/1",
            "avg_frame_rate": "25/1",
            "time_base": "1/12800",
            "start_pts": 0,
            "start_time": "0.000000",
            "duration_ts": 512,
            "duration": "0.040000",
            "bit_rate": "324800",
            "nb_frames": "1",
            "extradata_size": 2405,
            "disposition": {
                "default": 1,
                "dub": 0,
                "original": 0,
                "comment": 0,
                "lyrics": 0,
                "karaoke": 0,
                "forced": 0,
                "hearing_impaired": 0,
                "visual_impaired": 0,
                "clean_effects": 0,
                "attached_pic": 0,
                "timed_thumbnails": 0,
                "captions": 0,
                "descriptions": 0,
                "metadata": 0,
                "dependent": 0,
                "still_image": 0
            },
            "tags": {
                "language": "und",
                "handler_name": "VideoHandler",
                "vendor_id": "[0][0][0][0]"
            }
        }
    ]
}
```

**Comment by wader** [comment]
> This is way above my head since I have no clew how to even go to troubleshoot that :-). All and eBay help will be highly appreciated.

No problem, i will digg into when i get time, first is to get myself some sane arm environment :) emulated docker builds have proved to be way too slow and a bit "leaky" emulation wise.

I suspect it has to do with `-Wl,-z,relro` and it would be nice to keep using that for security, maybe can do exceptions somehow but that is probably messy also. I noticed that some of the aarch64 assembly in x265 use ro sections so maybe it's related to that somehow

**Comment by wader** [comment]
Tiny bit of progress: It seems to build fine without `-static-pie`. `-DENABLE_PIC=ON` did not make any difference and i've tried to change some of the aarch64 assembly section that looked suspicious but didn't how, either not the issue and i missed something.

I will probably have to get a proper arm host somewhere to debug this. Know any good place to get instance cheaply? I tried AWS once but seemed to be a mess to just get shell access, but maybe i missing something? :)

Maybe could file an issue upstream about this but feel i need to know more what is going on first.

File: Dockerfile (hunk 4 above)
Comment (wader) [reply Dockerfile:673]: Aha i see, the multi depth build setup is a bit confusing. I wonder why not all depths are included by default? maybe opt-out instead?

**Comment by wader** [comment]
Some progress:

It seems like the link error with `-static-pie` is unrelated to multilib. I get the same error with master not building with multilib. I tried to hackily bisect down between 3.5 and master which commit made `-static-pie` break and it seem to be https://bitbucket.org/multicoreware/x265_git/commits/32b08c771c5fd0ec4ed7adf952e9866b96791e38

**Comment by wader** [comment]
> There might be some clues in that patch if it is working as expected.

Oh thanks had forgot about them. Did a quick peek at the patch now and interestingly it's quite different from the x265 master arm4 code. Its all C with arm64 intrinsic instead of arm64 assembly, i wonder how much they differ in performance. Gut feeling is it will probably link fine with `-static-pie` but feels like a dead end code wise if x265 went their own way?

**Comment by wader** [comment]
I think we cracked it, i noticed now that x86 sets `-DPIC` in `ASM_FLAGS` in addition to `-fPIC` which arm64 also has and i assume was controller `defined(PIC)` in asm code, but turn out not. With this patch for the commit where it started failing ffmpeg configure is now happy:
```diff
diff --git a/source/CMakeLists.txt b/source/CMakeLists.txt
index 60031df63..9689b85af 100755
--- a/source/CMakeLists.txt
+++ b/source/CMakeLists.txt
@@ -260,7 +260,7 @@ if(GCC)
         endif()
     endif()
 	if(ARM64 OR CROSS_COMPILE_ARM64)
-	    set(ARM_ARGS -fPIC -flax-vector-conversions)
+	    set(ARM_ARGS -DPIC -fPIC -flax-vector-conversions)
         find_package(Neon)
         if(CPU_HAS_NEON)
             add_definitions(-DHAVE_NEON)
```

Could you do a hacky-sed thingy for the PR and we try? if it works we should probably file a upstream issue about it... also then a bit curious how that code works in a PIE?

**Comment by wader** [comment]
Also while i remember: Maybe should add a note to the readme that x265 is built multilib?

**Comment by wader** [comment]
> Nice. I was about to suggest to try compiling ffmpeg with
> 
> ```diff
> --- a/Dockerfile
> +++ b/Dockerfile
> @@ -734,13 +734,15 @@ RUN \
>    sed -i 's/add_ldexeflags -fPIE -pie/add_ldexeflags -fPIE -static-pie/' configure && \
>    ./configure \
>    --pkg-config-flags="--static" \
> -  --extra-cflags="-fopenmp" \
> +  --extra-cflags="-fPIC -fopenmp" \
> ```
> 
> But probably that is not going to help since the problem is in x265. Good one!

As the resulting binary linked as PIE i think it should fail linking if some code is not PIC, but maybe that is only true for  C/C++ object files? a bit curious why -static-pic detect this, have feeling it a security feature that it does not allow update after linking for something and has sanity check

**Comment by wader** [comment]
> > @wader done, hope this works! (fingers crossed)
> 
> Wait, one error

Ok, pushed updated tag

**Comment by wader** [comment]
> This did it, arm64 is compiled and pushed now! 🥇

Yeap!

```
➜  ~ docker run -v "$PWD:$PWD" -w "$PWD" --rm mwader/static-ffmpeg:mutliplatform-test-add-10-12bit-x265-arm64 -y -v quiet -f lavfi -i testsrc -pix_fmt yuv444p12le -t 20ms -c:v libx265 test.mp4
WARNING: The requested image's platform (linux/arm64/v8) does not match the detected host platform (linux/amd64) and no specific platform was requested
x265 [info]: HEVC encoder version 3.5+1-f0c1022b6
x265 [info]: build info [Linux][GCC 11.2.1][32 bit] 12bit
x265 [info]: using cpu capabilities: NEON
x265 [warning]: halving the quality when psy-rd is enabled for 444 input. Setting cbQpOffset = 6 and crQpOffset = 6
x265 [info]: Main 4:4:4 12 profile, Level-2 (Main tier)
x265 [warning]: No thread pool allocated, --wpp disabled
x265 [warning]: No thread pool allocated, --lookahead-slices disabled
x265 [info]: Slices                              : 1
x265 [info]: frame threads / pool features       : 2 / none
x265 [info]: Coding QT: max CU size, min CU size : 64 / 8
x265 [info]: Residual QT: max TU size, max depth : 32 / 1 inter / 1 intra
x265 [info]: ME / range / subpel / merge         : hex / 57 / 2 / 3
x265 [info]: Keyframe min / max / scenecut / bias  : 25 / 250 / 40 / 5.00
x265 [info]: Cb/Cr QP Offset                     : 6 / 6
x265 [info]: Lookahead / bframes / badapt        : 20 / 4 / 2
x265 [info]: b-pyramid / weightp / weightb       : 1 / 1 / 0
x265 [info]: References / ref-limit  cu / depth  : 3 / off / on
x265 [info]: AQ: mode / str / qg-size / cu-tree  : 2 / 1.0 / 32 / 1
x265 [info]: Rate Control / qCompress            : CRF-28.0 / 0.60
x265 [info]: tools: rd=3 psy-rd=2.00 early-skip rskip mode=1 signhide tmvp
x265 [info]: tools: b-intra strong-intra-smoothing deblock sao
x265 [info]: frame I:      1, Avg QP:32.74  kb/s: 319.00
x265 [info]: consecutive B-frames: 100.0% 0.0% 0.0% 0.0% 0.0%

encoded 1 frames in 4.15s (0.24 fps), 319.00 kb/s, Avg QP:32.74
➜  ~ docker run -v "$PWD:$PWD" -w "$PWD" --rm --entrypoint=/ffprobe mwader/static-ffmpeg:mutliplatform-test-add-10-12bit-x265-arm64 -v quiet -of json -show_streams -i test.mp4 | jq -r '.streams[0].pix_fmt'
WARNING: The requested image's platform (linux/arm64/v8) does not match the detected host platform (linux/amd64) and no specific platform was requested
yuv444p12le
```

**Comment by wader** [comment]
> Ok, could you please let me know the following:
> 
> * [ ]  How do you want to apply the patch?

Let's do sed for now and i can create an upstream issue, hopefully we can cleanup it up soon. Maybe add some comment, i can add a link to the issue also.

> * [ ]  (me) Update the readme to mention multilib (10+ bit) support in x265

Yeap do that

> * [ ]  Check the bump section of the Dockerfile to see if it is correct

Will do tomorrow

> * [ ]  (me) probably keep commits for README and Doceker updates separately

Yes. I update docker hub README manually at the moment

**Comment by wader** [comment]
Adding a test is only straight forward atm if ffmpeg exits with non-zero. Maybe skip test for this PR i can have a look later

File: Dockerfile [hunk 5]
```diff
@@ -653,30 +653,26 @@ RUN \
   ./configure --enable-pic --enable-static --disable-cli --disable-lavf --disable-swscale && \
   make -j$(nproc) install
 
-# bump: x265 /X265_VERSION=([\d.]+)/ https://bitbucket.org/multicoreware/x265_git.git|^3
+# x265 release is over 1 years old and master branch has a lot of fixes and improvements, so we checkout commit so no hash is needed
+# bump: x264 /X264_VERSION=([[:xdigit:]]+)/ gitrefs:https://bitbucket.org/multicoreware/x265_git.git|re:#^refs/heads/master$#|@commit
```
Comment (wader) [inline Dockerfile:657]: Change to `bump: x265 /X265_VERSION=([[:xdigit:]]+)/ gitrefs:https://bitbucket.org/multicoreware/x265_git.git|re:#^refs/heads/master$#|@commit`

File: Dockerfile [hunk 6]
```diff
@@ -653,30 +653,26 @@ RUN \
...
-    -DENABLE_NASM=ON \
-    -DCMAKE_BUILD_TYPE=Release \
-    ../../source && \
-  make -j$(nproc) install
+# TODO: remove 'sed' hack when upstream (x265) fixes the issue and adds '-DPIC' to ARM_ARGS
+RUN \
+  wget $WGET_OPTS -O x265_git.tar.bz2 "$X265_URL" && \
+  tar xf x265_git.tar.bz2 && \
+  cd multicoreware-x265_git-*/build/linux && \
+  sed -i '/^cmake / s/$/ -G "Unix Makefiles" ${CMAKEFLAGS}/' ./multilib.sh && \
```
Comment (wader) [inline Dockerfile:669]: Filed an issue about this also https://bitbucket.org/multicoreware/x265_git/issues/620/support-passing-cmake-flags-to-multilibsh

File: Dockerfile [hunk 7]
```diff
@@ -653,30 +653,26 @@ RUN \
...
+# x265 release is over 1 years old and master branch has a lot of fixes and improvements, so we checkout commit so no hash is needed
+# bump: x264 /X264_VERSION=([[:xdigit:]]+)/ gitrefs:https://bitbucket.org/multicoreware/x265_git.git|re:#^refs/heads/master$#|@commit
 # bump: x265 after ./hashupdate Dockerfile X265 $LATEST
-# bump: x265 link "releasenotes" https://bitbucket.org/multicoreware/x265_git/src/master/doc/reST/releasenotes.rst
-ARG X265_VERSION=3.5
-ARG X265_URL="https://bitbucket.org/multicoreware/x265_git/downloads/x265_$X265_VERSION.tar.gz"
-ARG X265_SHA256=e70a3335cacacbba0b3a20ec6fecd6783932288ebc8163ad74bcc9606477cae8
+# bump: x265 link "Source diff $CURRENT..$LATEST" https://bitbucket.org/multicoreware/x265_git/branches/compare/$LATEST..$CURRENT#diff
+ARG X265_VERSION=931178347b3f73e40798fd5180209654536bbaa5
+ARG X265_URL="https://bitbucket.org/multicoreware/x265_git/get/${X265_VERSION}.tar.bz2"
```
[the remark this answers is not in the document]
Comment (wader) [reply Dockerfile:661]: Aha i see hmm, let me have a look, it's very comfortable when bump can do all the changes needed

[the remark this answers is not in the document]
Comment (wader) [reply Dockerfile:661]: Think we can do this (not sure if based on your latest changes):

```Dockerfile
# x265 release is over 1 years old and master branch has a lot of fixes and improvements, so we checkout commit so no hash is needed
# bump: x265 /X265_VERSION=([[:xdigit:]]+)/ gitrefs:https://bitbucket.org/multicoreware/x265_git.git|re:#^refs/heads/master$#|@commit
# bump: x265 after ./hashupdate Dockerfile X265 $LATEST
# bump: x265 link "Source diff $CURRENT..$LATEST" https://bitbucket.org/multicoreware/x265_git/branches/compare/$LATEST..$CURRENT#diff
ARG X265_VERSION=931178347b3f73e40798fd5180209654536bbaa5
ARG X265_SHA256=eb07c919a62d4a6724b44427844c214ab7ddf57c85b3ccf3f2b8111763ac2966
ARG X265_URL="https://bitbucket.org/multicoreware/x265_git/get/$X265_VERSION.tar.bz2"
# -w-macro-params-legacy to not log lots of asm warnings
# https://bitbucket.org/multicoreware/x265_git/issues/559/warnings-when-assembling-with-nasm-215
# TODO: remove 'sed' hack when upstream (x265) fixes the issue and adds '-DPIC' to ARM_ARGS
RUN \
  wget $WGET_OPTS -O x265_git.tar.bz2 "$X265_URL" && \
  echo "$X265_SHA256  x265_git.tar.bz2" | sha256sum --status -c - && \
  tar xf x265_git.tar.bz2 && \
  cd multicoreware-x265_git-*/build/linux && \
  sed -i '/^cmake / s/$/ -G "Unix Makefiles" ${CMAKEFLAGS}/' ./multilib.sh && \
  sed -i 's/ -DENABLE_SHARED=OFF//g' ./multilib.sh && \
  sed -i 's/set(ARM_ARGS -fPIC -flax-vector-conversions)/set(ARM_ARGS -DPIC -fPIC -flax-vector-conversions)/' ../../source/CMakeLists.txt && \
  MAKEFLAGS="-j$(nproc)" \
  CMAKEFLAGS="-DENABLE_SHARED=OFF -DCMAKE_VERBOSE_MAKEFILE=ON -DENABLE_AGGRESSIVE_CHECKS=ON -DCMAKE_ASM_NASM_FLAGS=-w-macro-params-legacy -DENABLE_NASM=ON -DCMAKE_BUILD_TYPE=Release" \
  ./multilib.sh && \
  make -C 8bit -j$(nproc) install
```

Worked when i tried locally (`bump -i x265 update -r -v`) and docker build could download and verify

**Comment by wader** [comment]
🥳 thanks for all the help!

**Comment by wader** [comment]
Fixed upstream now https://github.com/wader/static-ffmpeg/pull/294 commit https://bitbucket.org/multicoreware/x265_git/commits/02a961bcd26cb299108a2065692500ab895059e9
